nifi-nar-bundles/nifi-aws-bundle/nifi-aws-abstract-processors/src/main/java/org/apache/nifi/processors/aws/v2/AbstractAwsProcessor.java

```diff
@@ -325,11 +344,11 @@ protected T getClient(final ProcessContext context) {
      */
     protected abstract U createClientBuilder(final ProcessContext context);
 
-    protected Region getRegion(final ProcessContext context) {
+    protected Region getRegion(final ProcessContext context, final FlowFile flowfile) {
         final Region region;
         // if the processor supports REGION, get the configured region.
         if (getSupportedPropertyDescriptors().contains(REGION)) {
-            final String regionValue = context.getProperty(REGION).getValue();
+            final String regionValue = context.getProperty(REGION).evaluateAttributeExpressions(flowfile).getValue();
```

Comment:
@grishick Thank you for opening this PR for improving the Region configuration in AWS processors!

Could you please explain your use case and flow configuration in more detail?

What I don't understand is that `REGION` is a property with `AllowableValue`-s and therefore it does not / can not support Expression Language. So calling `evaluateAttributeExpressions(flowfile)` has no effect at all.
What FlowFile attribute are you using in your flow? How do you configure this attribute for the AWS processor to use?

Reply:
@turcsanyip hm... from what I can see `evaluateAttributeExpressions` works just fine in support/nifi-1.x branch in our fork here: https://github.com/Eng-Anetac/nifi. In our use case, we deploy the same process group template from Nifi Registry to multiple NiFi instances across the world (usually one deployment per customer). In these process groups we use ListS3, FetchS3Object, PutS3Object, PutDynamoDB and other processors. When we deploy the process group, we also create a parameter context with parameter that specifies the AWS region. All AWS processors use that parameter to set the AWS region. Here is a snippet from one of our process groups. In this example #{customer_region} is the parameter:
```
{
        "autoTerminatedRelationships": [],
        "backoffMechanism": "PENALIZE_FLOWFILE",
        "bulletinLevel": "WARN",
        "bundle": {
          "artifact": "nifi-aws-nar",
          "group": "org.apache.nifi",
          "version": "1.27.0-ANETAC"
        },
        "comments": "",
        "componentType": "PROCESSOR",
        "concurrentlySchedulableTaskCount": 1,
        "executionNode": "PRIMARY",
        "groupIdentifier": "ad3652ba-9259-3356-8bba-058bd3f72c23",
        "identifier": "4f98e6dc-a48a-3741-b058-d3dd1790c255",
        "instanceIdentifier": "1631adbb-271e-3228-ddf0-97cc3a847702",
        "maxBackoffPeriod": "10 mins",
        "name": "List inventory folder",
        "penaltyDuration": "30 sec",
        "position": {
          "x": 4696.0,
          "y": -344.0
        },
        "properties": {
          "list-type": "1",
          "et-initial-listing-target": "all",
          "prefix": "inventory/#{domain_name}",
          "et-time-window": "3 hours",
          "requester-pays": "false",
          "listing-strategy": "timestamps",
          "Signer Override": "Default Signature",
          "min-age": "0 sec",
          "write-s3-user-metadata": "false",
          "max-age": "#{max_file_age}",
          "AWS Credentials Provider service": "#{AWSCredentialsProvider}",
          "Listing Batch Size": "#{min_batch_size}",
          "write-s3-object-tags": "false",
          "Bucket": "#{bucket_name}",
          "Communications Timeout": "30 secs",
          "Region": "#{customer_region}",
          "use-versions": "false"
        },
        "propertyDescriptors": {
          "Proxy Host": {
            "displayName": "Proxy Host",
            "identifiesControllerService": false,
            "name": "Proxy Host",
            "sensitive": false
          },
          "list-type": {
            "displayName": "List Type",
            "identifiesControllerService": false,
            "name": "list-type",
            "sensitive": false
          },
          "et-initial-listing-target": {
            "displayName": "Entity Tracking Initial Listing Target",
            "identifiesControllerService": false,
            "name": "et-initial-listing-target",
            "sensitive": false
          },
          "Access Key": {
            "displayName": "Access Key ID",
            "identifiesControllerService": false,
            "name": "Access Key",
            "sensitive": true
          },
          "prefix": {
            "displayName": "Prefix",
            "identifiesControllerService": false,
            "name": "prefix",
            "sensitive": false
          },
          "record-writer": {
            "displayName": "Record Writer",
            "identifiesControllerService": true,
            "name": "record-writer",
            "sensitive": false
          },
          "proxy-configuration-service": {
            "displayName": "Proxy Configuration Service",
            "identifiesControllerService": true,
            "name": "proxy-configuration-service",
            "sensitive": false
          },
          "Endpoint Override URL": {
            "displayName": "Endpoint Override URL",
            "identifiesControllerService": false,
            "name": "Endpoint Override URL",
            "sensitive": false
          },
          "et-time-window": {
            "displayName": "Entity Tracking Time Window",
            "identifiesControllerService": false,
            "name": "et-time-window",
            "sensitive": false
          },
          "custom-signer-class-name": {
            "displayName": "Custom Signer Class Name",
            "identifiesControllerService": false,
            "name": "custom-signer-class-name",
            "sensitive": false
          },
          "requester-pays": {
            "displayName": "Requester Pays",
            "identifiesControllerService": false,
            "name": "requester-pays",
            "sensitive": false
          },
          "custom-signer-module-location": {
            "displayName": "Custom Signer Module Location",
            "identifiesControllerService": false,
            "name": "custom-signer-module-location",
            "resourceDefinition": {
              "cardinality": "MULTIPLE",
              "resourceTypes": [
                "FILE",
                "DIRECTORY"
              ]
            },
            "sensitive": false
          },
          "listing-strategy": {
            "displayName": "Listing Strategy",
            "identifiesControllerService": false,
            "name": "listing-strategy",
            "sensitive": false
          },
          "Signer Override": {
            "displayName": "Signer Override",
            "identifiesControllerService": false,
            "name": "Signer Override",
            "sensitive": false
          },
          "delimiter": {
            "displayName": "Delimiter",
            "identifiesControllerService": false,
            "name": "delimiter",
            "sensitive": false
          },
          "proxy-user-password": {
            "displayName": "Proxy Password",
            "identifiesControllerService": false,
            "name": "proxy-user-password",
            "sensitive": true
          },
          "Secret Key": {
            "displayName": "Secret Access Key",
            "identifiesControllerService": false,
            "name": "Secret Key",
            "sensitive": true
          },
          "min-age": {
            "displayName": "Minimum Object Age",
            "identifiesControllerService": false,
            "name": "min-age",
            "sensitive": false
          },
          "write-s3-user-metadata": {
            "displayName": "Write User Metadata",
            "identifiesControllerService": false,
            "name": "write-s3-user-metadata",
            "sensitive": false
          },
          "max-age": {
            "displayName": "Maximum Object Age",
            "identifiesControllerService": false,
            "name": "max-age",
            "sensitive": false
          },
          "Listing Batch Size": {
            "displayName": "Listing Batch Size",
            "identifiesControllerService": false,
            "name": "Listing Batch Size",
            "sensitive": false
          },
          "SSL Context Service": {
            "displayName": "SSL Context Service",
            "identifiesControllerService": true,
            "name": "SSL Context Service",
            "sensitive": false
          },
          "write-s3-object-tags": {
            "displayName": "Write Object Tags",
            "identifiesControllerService": false,
            "name": "write-s3-object-tags",
            "sensitive": false
          },
          "Bucket": {
            "displayName": "Bucket",
            "identifiesControllerService": false,
            "name": "Bucket",
            "sensitive": false
          },
          "Communications Timeout": {
            "displayName": "Communications Timeout",
            "identifiesControllerService": false,
            "name": "Communications Timeout",
            "sensitive": false
          },
          "Region": {
            "displayName": "Region",
            "identifiesControllerService": false,
            "name": "Region",
            "sensitive": false
          },
          "Credentials File": {
            "displayName": "Credentials File",
            "identifiesControllerService": false,
            "name": "Credentials File",
            "resourceDefinition": {
              "cardinality": "SINGLE",
              "resourceTypes": [
                "FILE"
              ]
            },
            "sensitive": false
          },
          "et-state-cache": {
            "displayName": "Entity Tracking State Cache",
            "identifiesControllerService": true,
            "name": "et-state-cache",
            "sensitive": false
          },
          "proxy-user-name": {
            "displayName": "Proxy Username",
            "identifiesControllerService": false,
            "name": "proxy-user-name",
            "sensitive": false
          },
          "Proxy Host Port": {
            "displayName": "Proxy Host Port",
            "identifiesControllerService": false,
            "name": "Proxy Host Port",
            "sensitive": false
          },
          "use-versions": {
            "displayName": "Use Versions",
            "identifiesControllerService": false,
            "name": "use-versions",
            "sensitive": false
          }
        },
        "retriedRelationships": [],
        "retryCount": 10,
        "runDurationMillis": 0,
        "scheduledState": "ENABLED",
        "schedulingPeriod": "#{poll_frequency}",
        "schedulingStrategy": "TIMER_DRIVEN",
        "style": {},
        "type": "org.apache.nifi.processors.aws.s3.ListS3",
        "yieldDuration": "1 sec"
      },
```

Reply:
@grishick Thanks for the details!

So you are using parameters from Parameter Context, not Expression Language referencing FlowFile attributes. It makes a difference.
Parameters are evaluated automatically by the framework. No component level code is needed like calling `evaluateAttributeExpressions()`. The FlowFile is also not needed, because parameters come from Parameter Contexts, not from FlowFiles.

Configuring AWS region via parameters works in vanilla Apache NiFi too. The change in this PR is not needed for that.
Actually, this PR affects `org.apache.nifi.processors.aws.v2.AbstractAwsProcessor` which is not the parent of the S3 processors. S3 processors come from `org.apache.nifi.processors.aws.AbstractAWSProcessor` (this is the abstract processor using the older AWS client SDK, S3 processors have not yet been migrated to the v2 client lib). `AbstractAWSProcessor` [simply gets the Region property](https://github.com/Eng-Anetac/nifi/blob/4d33c35688004b6e00a06ad24276c72cb9027b73/nifi-nar-bundles/nifi-aws-bundle/nifi-aws-abstract-processors/src/main/java/org/apache/nifi/processors/aws/AbstractAWSProcessor.java#L363) in your fork and the S3 processors properly work for you with `#{ad_s3_region}` parameter reference. It proves that adding `evaluateAttributeExpressions(FlowFile)` is unnecessary.

Reply:
You are correct. I seem to have been running in circles with other features and this ended up being unnecessary for this particular use case. I had other processors that I wrote (e.g. listing EC2 instances) and those needed to be able to take the region from a flowfile attribute. But obviously, this does not work for the existing processors without further changes.